prov/rxm/src/rxm_ep.c

```diff
@@ -1477,62 +1492,76 @@ rxm_ep_send_common(struct rxm_ep *rxm_ep, const struct iovec *iov, void **desc,
-				      &tx_entry->rx_buf->recv_entry->rndv.tx_buf->pkt,
-				      tx_entry->deferred_pkt_size,
-				      tx_entry->rx_buf->recv_entry->rndv.tx_buf->hdr.desc,
-				      0, tx_entry->rx_buf);
+		def_tx_entry = container_of(rxm_conn->deferred_tx_queue.next,
+					    struct rxm_deferred_tx_entry, entry);
+		switch (def_tx_entry->type) {
+		case RXM_DEFERRED_TX_RNDV_ACK:
+			ret = fi_send(def_tx_entry->rxm_conn->msg_ep,
+				      &def_tx_entry->rndv_ack.rx_buf->
+					recv_entry->rndv.tx_buf->pkt,
+				      sizeof(def_tx_entry->rndv_ack.rx_buf->
+					recv_entry->rndv.tx_buf->pkt),
+				      def_tx_entry->rndv_ack.rx_buf->recv_entry->
+					rndv.tx_buf->hdr.desc,
+				      0, def_tx_entry->rndv_ack.rx_buf);
 			if (OFI_UNLIKELY(ret)) {
 				if (OFI_LIKELY(ret == -FI_EAGAIN))
 					break;
-				rxm_cq_write_error(tx_entry->ep->util_ep.rx_cq,
-						   tx_entry->ep->util_ep.rx_cntr,
-						   tx_entry->context, ret);
+				rxm_cq_write_error(def_tx_entry->rxm_ep->util_ep.rx_cq,
+						   def_tx_entry->rxm_ep->util_ep.rx_cntr,
+						   def_tx_entry->rndv_read.rx_buf->
+							recv_entry->context, ret);
 			}
-			rxm_ep_dequeue_deferred_tx_queue(tx_entry);
-			free(tx_entry);
+			rxm_ep_dequeue_deferred_tx_queue(def_tx_entry);
+			free(def_tx_entry);
 			break;
-		case RXM_RNDV_READ:
-			ret = fi_readv(tx_entry->conn->msg_ep,
-				       tx_entry->rma_buf->def.rxm_iov.iov,
-				       tx_entry->rma_buf->def.rxm_iov.desc,
-				       tx_entry->rma_buf->def.rxm_iov.count, 0,
-				       tx_entry->rma_buf->def.rxm_rma_iov.iov[0].addr,
-				       tx_entry->rma_buf->def.rxm_rma_iov.iov[0].key,
-				       tx_entry->rx_buf);
+		case RXM_DEFERRED_TX_RNDV_READ:
+			ret = fi_readv(def_tx_entry->rxm_conn->msg_ep,
+				       def_tx_entry->rndv_read.rxm_iov.iov,
+				       def_tx_entry->rndv_read.rxm_iov.desc,
+				       def_tx_entry->rndv_read.rxm_iov.count, 0,
+				       def_tx_entry->rndv_read.rma_iov.addr,
+				       def_tx_entry->rndv_read.rma_iov.key,
+				       def_tx_entry->rndv_read.rx_buf);
 			if (OFI_UNLIKELY(ret)) {
 				if (OFI_LIKELY(ret == -FI_EAGAIN))
 					break;
-				rxm_cq_write_error(tx_entry->ep->util_ep.rx_cq,
-						   tx_entry->ep->util_ep.rx_cntr,
-						   tx_entry->context, ret);
+				rxm_cq_write_error(def_tx_entry->rxm_ep->util_ep.rx_cq,
+						   def_tx_entry->rxm_ep->util_ep.rx_cntr,
+						   def_tx_entry->rndv_read.rx_buf->
+							recv_entry->context, ret);
 				break;
 			}
-			rxm_ep_dequeue_deferred_tx_queue(tx_entry);
-			free(tx_entry->rma_buf);
-			free(tx_entry);
-			break;
-		case RXM_SAR_TX:	/* SAR (TX segments) */
-			ret = rxm_ep_progress_sar_deferred_tx_queue(tx_entry);
+			rxm_ep_dequeue_deferred_tx_queue(def_tx_entry);
+			free(def_tx_entry);
 			break;
-		default:
-			FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
-				"The deferred operation (TX state - %d) "
-				"doesn't have registered hanlder\n",
-				tx_entry->state);
-			ret = -FI_EAGAIN;
-			assert(0);
+		case RXM_DEFERRED_TX_SAR_SEG:
```

Comment:
Why do we need new states?

Reply:
This is type of deferred operation, because we need to cast `rxm_deferred_tx_entry` to an appropriate TX buffer

Reply:
Please add why this commit is needed in the commit message.

Reply:
The reason of this commit has been added to the commit message